main.c

```diff
@@ -30,30 +43,147 @@ static void lv_linux_disp_init(void)
 #elif LV_USE_SDL
 static void lv_linux_disp_init(void)
 {
-    const int width = atoi(getenv("LV_SDL_VIDEO_WIDTH") ? : "800");
-    const int height = atoi(getenv("LV_SDL_VIDEO_HEIGHT") ? : "480");
 
-    lv_sdl_window_create(width, height);
+    lv_sdl_window_create(window_width, window_height);
+
+}
+#elif LV_USE_WAYLAND
+static void lv_linux_disp_init(void)
+{
+    lv_display_t *disp;
+    lv_group_t *g;
+
+    lv_wayland_init();
+    disp = lv_wayland_window_create(window_width, window_height, "LVGL Simulator", NULL);
+
+    if (fullscreen) {
+            lv_wayland_window_set_fullscreen(disp, fullscreen);
+    } else if (maximize) {
+            lv_wayland_window_set_maximized(disp, maximize);
+    }
+
+    g = lv_group_create();
+    lv_group_set_default(g);
+    lv_indev_set_group(lv_wayland_get_keyboard(disp), g);
+    lv_indev_set_group(lv_wayland_get_pointeraxis(disp), g);
+
 }
 #else
 #error Unsupported configuration
 #endif
 
-int main(void)
+#if !LV_USE_WAYLAND
+
+void default_run_loop(void)
+{
+    /*Handle LVGL tasks*/
+    while(1) {
+        lv_timer_handler();
+        usleep(5000);
+    }
+}
+
+
+#else
+
+void wayland_run_loop(void)
+{
+    struct pollfd pfd;
+    int sleep_wait = 0;
+    uint32_t time_till_next;
+    uint8_t tickless = 0;
+
+    pfd.fd = lv_wayland_get_fd();
+    pfd.events = POLLIN;
+
+    /* Handle LVGL tasks */
+    while (1) {
+
+        if (tickless == 0 && sleep_wait > 0) {
+            lv_tick_inc(sleep_wait);
+        }
+
+        /* Handle any Wayland/LVGL timers/events */
+        time_till_next = lv_wayland_timer_handler();
+
+        /* Run until the last window closes */
+        if (!lv_wayland_window_is_open(NULL)) {
+            break;
+        }
+
+        /* Wait for something for an event to happen */
+        if (time_till_next == LV_NO_TIMER_READY) {
+            sleep_wait = -1;
+        } else if (time_till_next > INT_MAX) {
+            sleep_wait = INT_MAX;
+        } else {
+            sleep_wait = time_till_next;
+        }
+
+        while ((poll(&pfd, 1, sleep_wait) < 0) && (errno == EINTR));
```

Comment:
Can it be part of the Wayland driver?
It would be nice to keep using `lv_timer_handler` as it is. Can we have a timer are theard in the driver for polling?

Reply:
``lv_wayland_timer_handler`` was introduced by previous contributors, to poll for events, if there are no events the application sleeps until the next refresh. I don't see how it's possible to do this in the driver, the call to ``poll(2)`` needs to know the time to next flush so it must be called from the main run loop.

Reply:
I am open for suggestions, but it's going to take some more time to refactor and re-test

Reply:
At first I tried disabling the event driven timer handling, to move out the call to ``lv_timer_handler`` out of the driver. But  early window maximization/fullscreen doesn't work anymore on weston... It crashes because the shared memory buffers are resized and thus re-arranged. So the references in the draw units of weston become incorrect and it crashes... This is the problem I had a few months back. And the solution was to build fixes around the event driven timer handler, because it more control over when the lvgl flush occurs. It can not occur in the middle of a commit. To add on top of it the event driver timer was a 'bit of hack' to beging with https://github.com/lvgl/lv_drivers/issues/207.  I will study the QNX driver more carefully, I see that they are pausing the timer. Will come up with something similar and make it run on weston...

Reply:
After the import, it's also probably a good to start a re-write, it needs it...

Reply:
@kisvegabor I've adjusted the driver as much as I could to match the other ones https://github.com/lvgl/lvgl/pull/6549/commits/7783303f704c8ac444dc414a8d2738a5792a8ca3
I prefer to implement the sleeping when the window hidden or minimized, once we refactor the driver. It currently works well on weston and Mutter which is great. Is that fine with you ? If yes, I will take of the stubs this evening/tommorow

Reply:
Re-introduced the sleep when hidden feature, by using the frame done event. It's a slightly different approach than original. Works well, tested and documented.